winsup/utils/getprocaddr.c

```diff
@@ -0,0 +1,198 @@
+
+  if (!SymFromAddrW (process, (DWORD64)(intptr_t)address, &displacement, info))
+    {
+      output (1, "Could not get symbol info\n");
+      SymCleanup(process);
+      return FALSE;
+    }
+
+  if (do_print_address)
+  {
+    output(0, "%p\n", (void *)(intptr_t)info->Address);
+  }
+  else
+  {
+    HANDLE main_process = OpenProcess(PROCESS_CREATE_THREAD |
+                                          PROCESS_QUERY_INFORMATION |
+                                          PROCESS_VM_OPERATION | PROCESS_VM_WRITE |
+                                          PROCESS_VM_READ,
+                                      FALSE, pid);
+    if (main_process == NULL)
+      return 1;
+    /* Inject the remote thread only when asked too*/
+    inject_remote_thread_into_process(main_process, (LPTHREAD_START_ROUTINE)(intptr_t)info->Address, exit_code);
+  }
+  CloseHandle(GetStdHandle(STD_OUTPUT_HANDLE));
+  SymCleanup(process);
+
+  exit(0);
+}
+
+int main(int argc, char **argv)
+{
+    char *end;
+    if (argc != 2 && argc < 4)
+    {
+        output(1, "Need a function name, exit code and pid\n");
+        output(1, "Or needs a function name.\n");
+        return 1;
+    }
+    if (argc == 4)
+    {
+        exit_code = atoi(argv[2]);
+        pid = strtoul(argv[3], NULL, 0);
+    }
+    if (argc == 2)
+    {
+        do_print_address = TRUE;
+    }
+
+    if (strcmp(argv[1], "CtrlRoutine"))
+    {
+        if (argc > 4)
+        {
+            output(1, "Unhandled option: %s\n", argv[4]);
+            return 1;
+        }
+
+        HINSTANCE kernel32 = GetModuleHandle("kernel32");
+        if (!kernel32)
+            return 1;
+        void *address = (void *)GetProcAddress(kernel32, argv[1]);
+        if (!address)
+            return 1;
+        if (do_print_address)
+        {
+            output(0, "%p\n", address);
+            return 0;
+        }
+        HANDLE main_process = OpenProcess(PROCESS_CREATE_THREAD |
+                                              PROCESS_QUERY_INFORMATION |
+                                              PROCESS_VM_OPERATION | PROCESS_VM_WRITE |
+                                              PROCESS_VM_READ,
+                                          FALSE, pid);
+        if (main_process == NULL)
+            return 1;
+        /* Inject the remote thread first*/
+        inject_remote_thread_into_process(main_process, (LPTHREAD_START_ROUTINE)address, exit_code);
+        return 0;
+    }
+    if (argc > 4)
+    {
+        output(1, "Unhandled option: %s\n", argv[4]);
+        return 1;
+    }
+
+    if (!SetConsoleCtrlHandler(ctrl_handler, TRUE))
+    {
+        output(1, "Could not register Ctrl handler\n");
+        return 1;
+    }
+    if (!GenerateConsoleCtrlEvent(CTRL_BREAK_EVENT, 0))
+    {
+        output(1, "Could not simulate Ctrl+Break\n");
+        return 1;
+    }
+
+    /* This is needed for both thread to inject
+    * and GenerateConsoleCtrlEvent.
+    * Gives 1 sec time. */
+    Sleep(1000);
```

Comment:
I don't like that `Sleep()` at all. It does not say "wait until the event has been processed".

A much better way would most likely be to use a semaphore. In Win32 land, that would entail calling `CreateEvent()`, storing the result in a global variable, calling `SetEvent()` in the `ctrl_handler` after everything is said and done, and calling `WaitForSingleObject()` here. See https://docs.microsoft.com/en-us/windows/win32/sync/using-event-objects for an example (look for `ghWriteEvent`).

Reply:
Looks good, I will give it a try later today.

Reply:
Ok done, https://github.com/msys2/msys2-runtime/pull/31/commits/29960882133de6aea3d350899eda9a182a8e2901